File: rename.pl
#!/usr/bin/perl

use 5.010;
use strict;
use warnings;
use IO::Uncompress::Unzip qw(unzip $UnzipError);
use Archive::Zip qw( :ERROR_CODES :CONSTANTS );use File::Copy qw(move);
use IO::Compress::Zip qw(zip $ZipError) ;
use List::MoreUtils;use Time::Piece;
use Archive::Extract;

open my $fh, ">", 'logfile.txt' 
  or die "Can't open the log file: $!";
  
  my $from_address = 'transportal@topslab.wisc.edu';
my $smtp_host = 'smtp.wiscmail.wisc.edu';
my $recipients = 'cheng8@wisc.edu';
my $fromEmailState = join "","From e-mail address is: ", $from_address, "\n";
my $smtpHostState = join "", "smtp host e-mail is ", $smtp_host, "\n";
my $recipientsState = join "", "Recipients e-mail is ", $recipients, "\n";
print $fh $fromEmailState;
print $fh $smtpHostState;
print $fh $recipientsState;
# Path to the directory
my $path = "C:/Users/Owner/Desktop/Perl";

# Open and read content of the directory
#opendir(D, "$path") || die "Can't open directory $path: $!\n";
if (opendir(D, "$path")) {
    print $fh "Open directory $path: $!\n";
} else {
    print $fh "Can't open directory $path: $!\n"; }
# Define the number of the past days to iterate
my $numPast = 2;
my $index = 1;
while ($index < $numPast) {

    # The date of last day to start 
    my $now = time();
    my $threeDaysAgo = $now - $index * 86400;
    my ($day, $mon, $year) = (localtime($threeDaysAgo))[3, 4, 5];
    printf("One day ago was %04d-%02d-%02d", $year+1900, $mon+1, $day);
    print "\n";
    my $dateState = join "","One day ago was ", $year+1900, "-", $mon+1, "-", $day, ".";
    print $fh $dateState;
    print $fh "\n";
    

    # Check file
    my $prefix = "d2d";
    my $yearStr = $year+1900;
    my $monStr = join "", "0", $mon+1;
    my $dayStr = join "", $day;
    my $dateString = join "", $yearStr, $monStr, $dayStr;
    my $postfix = ".zip";
    my $lastDayFile = join "", $prefix, $dateString, $postfix;
    print "Last Day '$lastDayFile' to start!\n";
    my $fileName = join "","Last Day ", $lastDayFile, " to start!";
    print $fh $fileName;
    print $fh "\n";    
        # Error Check
    if (-e $lastDayFile) { # Check whether file exists
        print "The last day file exists\n";
        print $fh "The last day file exists!\n";
    } else {
            print "The last day file '$lastDayFile' does not exist!\n";
            my $failState = join "","The last day file ", $lastDayFile, " does not exist!\n";
            print $fh $failState;
            print "Download the file '$lastDayFile'! \n";
            my $downloadState = join "", "Download the file", $lastDayFile, "!\n";
            print $fh $downloadState;
    }
        
    my $zip = Archive::Zip->new($lastDayFile);
    my $extractState = "Starting to extract!\n";
    print $fh $extractState;
    foreach my $member ($zip->members)
    {
     
        next if $member->isDirectory;
        (my $extractName = $member->fileName) =~ s{.*/}{};
        $member->extractToFileNamed(
        "$path/$extractName");
    }
    
        # Rename
    my $renameState = "Starting to rename!\n";
    print $fh $renameState;
    my $oldFile1 = join "", $prefix, $dateString, ".csv"; 
    my $oldFile2 = join "", $prefix, $dateString, "Controllers.csv";
    my $oldFile3 = join "", $prefix, $dateString, "Detectors.csv";
    my $newFile1 = "d2d_detlog.csv";
    my $newFile2 = "d2d_Controllers.csv";
    my $newFile3 = "d2d_Detectors.csv";
    move $oldFile1, $newFile1;
    move $oldFile2, $newFile2;
    move $oldFile3, $newFile3;
    
        # Delete Files
    my $deleteState = "Starting to delete!\n";
    print $fh $deleteState;
    unlink $newFile1;
    if(-e $newFile1) 
    {
        print "'$newFile1' file still exists!";
        my $newFileState1 = join "", $newFile1, " file stll exists!\n";
        print $fh $newFileState1;
    }
    else 
    {
        print "'$newFile1' file gone.";
        my $newFileGone1 = join "", $newFile1, " file gone!\n";
        print $fh $newFileGone1;
    }

    unlink $newFile2;
    if(-e $newFile2) 
    {
        print "'$newFile2' file still exists!";
        my $newFileState2 = join "", $newFile2, " file stll exists!\n";
        print $fh $newFileState2;
    }
    else 
    {
        print "'$newFile2' file gone.";
        my $newFileGone2 = join "", $newFile2, " file gone!\n";
        print $fh $newFileGone2;
    }

    unlink $newFile3;
    if(-e $newFile3) 
    {
        print "'$newFile3' file still exists!";
        my $newFileState3 = join "", $newFile3, " file stll exists!\n";
        print $fh $newFileState3;
    }
    else 
    {
        print "'$newFile3' file gone.";
        my $newFileGone3 = join "", $newFile3, " file gone!\n";
        print $fh $newFileGone3;
    }

    
    $index += 1;
    my $dayIndex = join "", "Iteration of day Index is ",$index,".\n";
    print $fh $dayIndex;}

File: rename.py
import numpy as np
#import pandas as pd
import os
import glob
from datetime import datetime, timedelta
import zipfile
from zipfile import ZipFile

# Define path of the directory
path = 'C:/Users/Owner/Desktop/Perl/'

# Define sender and receiver
sender = "transportal@topslab.wisc.edu"
host = "smtp.wiscmail.wisc.edu"
receiver = "cheng8@wisc.edu"

# Check current date and set the number of days ago
# Save the past N days *zip files into list
N = 1
check_file_list = []

while (N != 0):
    date_N_days_ago = datetime.now() - timedelta(days=N)
    file_N_days_ago = "d2d" + str(date_N_days_ago)[:4] + str(date_N_days_ago)[5:7] + str(date_N_days_ago)[8:10] + ".zip"
    check_file_list.append(file_N_days_ago)
    N -= 1
print(check_file_list)

# Create log file and write track information
with open('log.txt', 'w') as f:

    # Past N days file record
    f.write("Check files of " + str(N) + " days ago!\n")

    # Email information record
    f.write("Open log.txt!\n")
    f.write("Sent from: " + sender + "\n")
    f.write("Host: " + host + "\n")
    f.write("Receiver: " + receiver + "\n")

    # Open directory and iterate the *zip files
    f.write("Opening the *zip files in target directory " + path + "!\n") # Open the goal directory

    # Rename files
    file1 = "d2d_detlog.csv"
    file2 = "d2d_Controllers.csv"
    file3 = "d2d_Detectors.csv"


    for item in check_file_list:
        # Old file name
        old_file1 = item[:-4] + ".csv"
        old_file2 = item[:-4] + "Controllers.csv"
        old_file3 = item[:-4] + "Detectors.csv"

        fantasy_zip = zipfile.ZipFile(path + item)
        if(fantasy_zip):
            f.write("Starting to extract files!\n")
        else:
            f.write("Downloading to extract files!\n")

        fantasy_zip.extractall(path)
        fantasy_zip.close()

        os.rename(old_file1,file1)
        os.rename(old_file2, file2)
        os.rename(old_file3, file3)

        f.write("Rename " + item + " done!\n")
        os.remove(item)

        with ZipFile(item, 'w') as myzip:
            myzip.write(file1)
            myzip.write(file2)
            myzip.write(file3)

        myzip.close()

        os.remove(old_file1)
        os.remove(old_file2)
        os.remove(old_file3)




